### ISSUE TITLE
View table support?

### ISSUE THREAD

**Issue by itaied246** [sequelize/sequelize#7197]
Hello,

I'm using `sequelize` as my ORM for a small web app I'm writing as a hobby.
After modeling my data and creating the objects, I noticed that a view table can help me a lot querying data.
Searching the github repo, I found [this](https://github.com/sequelize/sequelize/issues/3078) question, and I wanted to know why isn't it possible to implement views on sequelize?

I have an `offer` model, which contains the following fields (uppercase name represents references for tables):
````
Offer
|--id
|--price
|__Game
|----name
|----platform
|____GiantGame
|--------image
````
All of my queries to offer will require the `image`, `platform` and `name`, so I need to execute a join (sequelize `include`).
I think a view here could make it much easier.
Is there another way to create the view and keep it consistent?

**Comment by dmlukichev** [comment]
Right now I'm having similar issue.

Indeed, views are not supported. 
But you can still use them after defining view's definition.
The only problem would be schema sync and this can be avoided by overriding view's model `sync()` and `drop()` methods.

```javascript
export default function (sequelize) {
    return sequelize.define('GameOfferView', {
        id: {
            type: Sequelize.STRING(32),
            primaryKey: true
        },
        price: Sequelize.REAL,
        platform: Sequelize.ENUM(),
        /* other columns */
    }, {
        tableName: 'game_offer_view',
        classMethods: {
            async sync(options) {
                const viewPath = pathResolve(__dirname, '/game-offer-view.sql');
                const viewSource = await promisify(readFile)(path, { encoding: 'utf-8' });

                await sequelize.query(viewSource, {
                    logging: options.logging,
                });
            },

            async drop(options) {
                return sequelize.query('DROP VIEW IF EXISTS game_offer_view', {
                    logging: options.logging,
                });
            }
        }
    });
}
```

If you have multiple views you can even add some utility function that takes model and adds that methods. If you will, I would advise you also to override `create()`, `update()` and `delete()` methods with something that throws errors. And publish this mixin so I could use it too :)

**Comment by phillippbertram** [comment]
This is not working (anymore?) with sequelize v4.37.1 :/ 

As the documentation says in [Upgrade-To-v4](http://docs.sequelizejs.com/manual/tutorial/upgrade-to-v4.html), `classMethods` and `instanceMethods` are removed. So tried the new way to define or override those methods like they are describing:

```javascript
const Model = sequelize.define('Model', {
    ...
});

// Class Method
Model.associate = function (models) {
    ...associate the models
};

// Instance Method
Model.prototype.someMethod = function () {..}
```
Here is what's my code looks like (simplified) in typesript

```typescript
model = return sequelize.define('GameOfferView', { ... });
model.sync = async function (options?: Sequelize.SyncOptions) {
    const viewPath = pathResolve(__dirname, '../../../sql/script_to_create_view.sql');
    const viewSource = await promisify(readFile)(viewPath, { encoding: 'utf-8' });
    await sequelize.query(viewSource, {
      logging: options.logging
    });
    return this as any;
 };
```

But I am not getting `sync` to be called, neither by using `classMethods` nor by using the `model.sync = ... ` way if I run `sequelize.sync({ force: true })`.

Maybe anyone else haas a solution?

Thanks in advance,
Phillipp

**Comment by thehappycoder** [comment]
I need this feature too.

**Comment by thehappycoder** [comment]
@phillippbertram 
Try to get rid of `async` and do something like this:
```
model.sync = (options: SyncOptions) => {
...
}
```

**Comment by abelosorio** [comment]
Not a perfect solution but it works well in Postgres

:point_right: https://github.com/abelosorio/sequelize-views-support

ref #3078

**Comment by LittleSaya** [comment]
My choice is writing and wrapping raw queries to mock the functionality of view,  though I still think it'd be better if there is an official support for it.

**Comment by papb** [comment]
@LittleSaya Have you tried [sequelize-views-support](https://github.com/abelosorio/sequelize-views-support)?

**Comment by LittleSaya** [comment]
@papb I'm using MySQL and sequlize-views-support's doc says it only supports PostgreSQL so I didn't try it.

**Comment by papb** [comment]
@LittleSaya Ah, I see! Ok.

**Comment by nichatr** [comment]
> My choice is writing and wrapping raw queries to mock the functionality of view, though I still think it'd be better if there is an official support for it.

Could you elaborate how it is done?
I'm using sequelize models for the tables and add associations, so the query for me is actually a findOne/findAll with the addition of "includes".

**Comment by amirh715** [comment]
Hello. Aren't views supported yet?

**Comment by wclawson** [comment]
I would really like postgres view support too

**Comment by paolocattani** [comment]
Any updates on views support?

**Comment by exos** [comment]
I'm working to add views support in a project where I'm working, I'm thinking publish it like a decorator, but if there are interest in this feature, I can send a PR to the main project, but, I need more time.

**Comment by tonypee** [comment]
@exos How did you go with this? Im interested in view support for mysql

**Comment by exos** [comment]
> @exos How did you go with this? Im interested in view support for mysql

Yes! sorry the delay, I was very bussy this time. I just publish this:

https://exoshtw.github.io/sequelize-view-support/

I did prepared this docuementation in some hours, I'm aware that this documentation is plagued with typos errors and misspelled text, sorry for that, my primary language are not the english. I wold appreciate for PR's fixing docuementation.

Thanks.

**Comment by Jayasankar-R** [comment]
import Sequelize from 'sequelize';
import { readFile } from 'fs';
import { promisify } from 'bluebird';
import { resolve as pathResolve } from 'path';


export default function defineGameOfferView(sequelize) {
    const GameOfferView = sequelize.define('GameOfferView', {
        id: {
            type: Sequelize.STRING(32),
            primaryKey: true
        },
        price: Sequelize.REAL,
        platform: Sequelize.ENUM(),
        /* other columns */
    }, {
        tableName: 'game_offer_view',
        classMethods: {
            // Override sync method to create the view
            async sync(options) {
                const viewPath = pathResolve(__dirname, '/game-offer-view.sql');
                const viewSource = await promisify(readFile)(viewPath, { encoding: 'utf-8' });

                await sequelize.query(viewSource, {
                    logging: options.logging,
                });
            },
            async drop(options) {
                return sequelize.query('DROP VIEW IF EXISTS game_offer_view', {
                    logging: options.logging,
                });
            },
            //  override create, update, and delete methods to throw errors
            async create(values, options) {
                throw new Error('Views cannot be directly modified');
            },
            async update(values, options) {
                throw new Error('Views cannot be directly modified');
            },
            async delete(options) {
                throw new Error('Views cannot be directly modified');
            }
        }
    });

    return GameOfferView;
}
